### ISSUE TITLE
HTML parser used for markup in translations breaks self-closing elements

### ISSUE THREAD

**Issue by stasm** [projectfluent/fluent.js#188]
Both `fluent-dom` and `fluent-react` currently use the `<template>` element to safely parse markup in translations. The `<template>` element internally uses the HTML parser. This is similar in effect to using `DOMParser.parseFromString` with `text/html` as supported type.

The HTML parser has rather strict rules on parsing void elements, unlike the XML parser. It only allows the [elements known to be void](https://html.spec.whatwg.org/multipage/syntax.html#void-elements) (e.g. `input`) to be self-closing (`<element/>`). Non-void elements are parsed as if they weren't properly closed.

This means that the following markup:

    A <span/> B

…is parsed as:

    A <span> B</span>

The consequence for overlays is that localizers can't use the self-closing form with non-void elements. Admittedly, they're not valid HTML markup, but the difference is rather subtle:

    # This inserts "!" into the span.
    hello = Hello, <span data-l10n-name="user" />!

    # This works as expected.
    hello = Hello, <span data-l10n-name="user"></span>!

**Comment by stasm** [comment]
One way to fix this would be to use `DOMParser` with the `application/xml` type for parsing markup in translations. XML has stricter parsing rules in general but it also allows any elements to be self-closing and it still produces well-formed documents.

We'd need to add the DTD for at least some HTML entities that localizers might expect to work. The [full DTD](https://www.w3.org/TR/html/entities.dtd) is huge, clocking in at over 11KB ([preview](https://dev.w3.org/html5/html-author/charref)). Perhaps we could only cherry pick and support whitespace characters, like `&nbsp;` and `&shy;`? In any case, it would be useful to strictly define the list of entities that we support.

**Comment by cvle** [comment]
I think I ran into this, I tried to do this:

```jsx
<Localized
  id="comments-userBoxAuthenticated-signedInAs"
  username={<Username />}
>
  <Typography variant="bodyCopy" container="div"> 
    {"Signed in as <username />."}
  </Typography>
</Localized>
```

```fluent
comments-userBoxAuthenticated-signedInAs =
  Signed in as <username/>.
```

But the `.` is swallowed at the end while this works:

```jsx
<Localized
  id="comments-userBoxAuthenticated-signedInAs"
  username={<Username />}
>
  <Typography variant="bodyCopy" container="div"> 
    {"Signed in as <username></username>."}
  </Typography>
</Localized>
```

```fluent
comments-userBoxAuthenticated-signedInAs =
  Signed in as <username></username>.
```

**Comment by stasm** [comment]
Yes, this looks familiar. I wrote a test to document the current (broken) behavior: https://github.com/projectfluent/fluent.js/blob/6e1ab3a4982ef6d5f385c77d15a16b27abc82a2e/fluent-react/test/localized_overlay_test.js#L618-L643

Do you think that using `<username></username>` is an acceptable work-around?

**Comment by stasm** [comment]
I'm not fond of it either. Perhaps using [`DOMParser`](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) and treating translations as XML rather than HTML, would work better? I'd need to check the performance impact, too.

**Comment by nchevobbe** [comment]
> I'm not fond of it either. Perhaps using [`DOMParser`](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) and treating translations as XML rather than HTML, would work better? I'd need to check the performance impact, too.

I think that would be a nice solution. 

We're using `fluent-react` in Firefox DevTools, and the document can be an xhtml one, and in such case custom tags are ignored:

```js
temp =  document.createElement("template")
temp.innerHTML = "<a>test</a>|<custom>testCustom</custom>"
Array.from(temp.content.childNodes) // [ a, #text, #text ]
```

whereas with `DOMParser` it works as expected:

```js
parser = new DOMParser()
doc = parser.parseFromString("<a>test</a>|<custom>testCustom</custom>", "text/html")
Array.from(doc.body.childNodes) // [ a, #text, custom ]
```

And sometimes we don't even have a `document` reference, which makes some of panel to fail loading when trying to update our version of fluent-react to `0.16.1`

**Comment by Pike** [comment]
Changing the parser logic can introduce subtle regressions, the HTML fixup is probably different between document and element.
And it needs a security review, as Stas mentioned.

I still like the lack of XML complexity in our DOM overlay stories.
